ebos/eclwriter.hh

```diff
@@ -207,13 +211,111 @@ public:
     /*!
      * \brief collect and pass data and pass it to eclIO writer
      */
-    void writeOutput(bool isSubStep)
+
+    void evalSummaryState(bool isSubStep)
     {
+        int reportStepNum = simulator_.episodeIndex() + 1;
+        /*
+          The summary data is not evaluated for timestep 0, that is
+          implemented with a:
+
+             if (time_step == 0)
+                 return;
+
+          check somewhere in the summary code. When the summary code was
+          split in separate methods Summary::eval() and
+          Summary::add_timestep() it was necessary to pull this test out
+          here to ensure that the well and group related keywords in the
+          restart file, like XWEL and XGRP were "correct" also in the
+          initial report step.
+
+          "Correct" in this context means unchanged behavior, might very
+          well be more correct to actually remove this if test.
+        */
+        if (reportStepNum == 0)
+            return;
+
+        const auto& summary = eclIO_->summary();
         Scalar curTime = simulator_.time() + simulator_.timeStepSize();
         Scalar totalCpuTime =
             simulator_.executionTimer().realTimeElapsed() +
             simulator_.setupTimer().realTimeElapsed() +
             simulator_.vanguard().externalSetupTime();
+
+        Opm::data::Wells localWellData = simulator_.problem().wellModel().wellData();
+
+        const auto& gridView = simulator_.vanguard().gridView();
+        int numElements = gridView.size(/*codim=*/0);
+        bool log = collectToIORank_.isIORank();
+        eclOutputModule_.allocBuffers(numElements, reportStepNum, isSubStep, log);
+
+        ElementContext elemCtx(simulator_);
+        ElementIterator elemIt = gridView.template begin</*codim=*/0>();
+        const ElementIterator& elemEndIt = gridView.template end</*codim=*/0>();
+        for (; elemIt != elemEndIt; ++elemIt) {
+            const Element& elem = *elemIt;
+            elemCtx.updatePrimaryStencil(elem);
+            elemCtx.updatePrimaryIntensiveQuantities(/*timeIdx=*/0);
+            eclOutputModule_.processElement(elemCtx);
```

Comment:
I do not see a corresponding call and loop in the master code, why is this necessary?

Reply:
I don't understand this code - but it is in master.

Reply:
I've spent some time in this area as part of the work on restart I/O.  Basically [`EclOutputBlackOilModule::processElement()`](https://github.com/OPM/opm-simulators/blob/e114d4a46d910ce0f18df6adff381668bbb1a2b3/ebos/ecloutputblackoilmodule.hh#L371-L375) extracts information that is maintained in the element context (e.g., cell pressure, saturations, gas/oil ratio, relative permeability) into linear buffers that are suitable for output to restart files (i.e., buffers that are used as backing data in subsequent calls to [`EclOutputBlackOilModule::assignToSolution()`](https://github.com/OPM/opm-simulators/blob/e114d4a46d910ce0f18df6adff381668bbb1a2b3/ebos/ecloutputblackoilmodule.hh#L779-L782) and [`EclOutputBlackOilModule::outputFipLog()`](https://github.com/OPM/opm-simulators/blob/e114d4a46d910ce0f18df6adff381668bbb1a2b3/ebos/ecloutputblackoilmodule.hh#L923-L924).)

Reply:
Thank you - yes that was my impression to, and since the summary output can potentially need this data for e.g. block pressure and fipxxx output I concluded - without checking very much, that it was needed even though the code in question does not output restart fields.

Reply:
Ok thanks.

I do not understand how to read this diff on github though: There are green lines with a plus for this code block, but no corresponding red lines with minus showing it was moved (try to search in the browser for processElement for example). However when I check the master, it is clearly there early in the writeOutput method. So there was one code block before, there is one now (net change zero), yet GitHub seems to show addition of the block with no corresponding removal (net change one). How can that be?

Reply:
> but no corresponding red lines with minus showing it was moved

Well it was not moved, it was copied.

> So there was one code block before [...]

and now there is two - the original in a code block with no changes.

Reply:
Thanks, I failed to spot the collapsed code part that contained the original block still. That brings up another question though: should it not be removed from its original spot?

Reply:
> That brings up another question though: should it not be removed from its original spot?

I saw that one coming; no it should not be removed from the original callsite. The point is that the code we are discussing now is used to serialize and broadcast the solution fields `PRESSURE, SWAT, ...` for output. The serialized solution fields are required both when evaluating the summary variables - the new callsite, and when storing restart files - the original callsite.

It might be possible to optimize a bit here, but I would like to merge "correct and working" first, and then possibly we can optimize a bit in the future.